Search/ReIndex/Provider/DoctrineOrmProvider.php

```diff
@@ -0,0 +1,122 @@
+
+use Doctrine\ORM\EntityManagerInterface;
+use Massive\Bundle\SearchBundle\Search\ReIndex\ReIndexProviderInterface;
+use Metadata\MetadataFactory;
+
+class DoctrineOrmProvider implements ReIndexProviderInterface
+{
+    /**
+     * @var ClassMetadataFactory
+     */
+    private $entityManager;
+
+    /**
+     * @var MetadataFactory
+     */
+    private $searchMetadataFactory;
+
+    /**
+     * @deprecated BC hack. @see this::provide().
+     *
+     * @var object[]
+     */
+    private $cachedEntities = [];
+
+    public function __construct(
+        EntityManagerInterface $entityManager,
+        MetadataFactory $searchMetadataFactory
+    ) {
+        $this->entityManager = $entityManager;
+        $this->searchMetadataFactory = $searchMetadataFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getClassFqns()
+    {
+        $metadataFactory = $this->entityManager->getMetadataFactory();
+        $classFqns = [];
+
+        foreach ($metadataFactory->getAllMetadata() as $classMetadata) {
+            if (null === $this->searchMetadataFactory->getMetadataForClass($classMetadata->name)) {
+                continue;
+            }
+            $classFqns[] = $classMetadata->name;
+        }
+
+        return $classFqns;
+    }
+
+    /**
+     * BC Note: Previous versions of the MassiveSearchBundle expected a collection of entities to
+     *          be returned from the repository via. a custom repository method. The expected behavior
+     *          now is that a query builder will be passed and NOTHING should be returned.
+     *
+     * {@inheritdoc}
+     */
+    public function provide($classFqn, $offset, $maxResults)
+    {
+        if (!empty($this->cachedEntities)) {
+            return $this->sliceEntities($offset, $maxResults);
+        }
+
+        $repository = $this->entityManager->getRepository($classFqn);
+        $metadata = $this->searchMetadataFactory->getMetadataForClass($classFqn);
+
+        $repositoryMethod = $metadata->getOutsideClassMetadata()->getReindexRepositoryMethod();
+
+        $queryBuilder = $repository->createQueryBuilder('d');
+
+        if ($repositoryMethod) {
+            $result = $repository->$repositoryMethod($queryBuilder);
+
+            if (is_array($result)) {
+                @trigger_error('Reindex repository methods should NOT return anything. Use the passed query builder instead.');
+                $this->cachedEntities = $result;
+
+                return $this->sliceEntities($offset, $maxResults);
+            }
+        }
+
+        $queryBuilder->setFirstResult($offset);
+        $queryBuilder->setMaxResults($maxResults);
+
+        return $queryBuilder->getQuery()->execute();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCount($classFqn)
+    {
+        $queryBuilder = $this->entityManager->createQueryBuilder();
+        $queryBuilder->select('count(a.id)');
+        $queryBuilder->from($classFqn, 'a');
```

Comment:
No it is not. But the issue is that we can't filter the seured documents
here without a large overhead. In the case where there are secured
documents, the user will finish "early" according to the progress
indicator (or we could "hide" this by always boostring the progress to
100% at the end.)

On Mon, Feb 22, 2016 at 05:50:37AM -0800, Daniel Rotter wrote:
>    In [1]Search/ReIndex/Provider/DoctrineOrmProvider.php:
> 
>  > +        }
>  > +
>  > +        $queryBuilder->setFirstResult($offset);
>  > +        $queryBuilder->setMaxResults($maxResults);
>  > +
>  > +        return $queryBuilder->getQuery()->execute();
>  > +    }
>  > +
>  > +    /**
>  > +     * {@inheritdoc}
>  > +     */
>  > +    public function getCount($classFqn)
>  > +    {
>  > +        $queryBuilder = $this->entityManager->createQueryBuilder();
>  > +        $queryBuilder->select('count(a.id)');
>  > +        $queryBuilder->from($classFqn, 'a');
> 
>    I think this is not true if secured entities are excluded, or is it?
> 
>    —
>    Reply to this email directly or [2]view it on GitHub.
> 
>    Reverse link: [3]unknown
> 
> References
> 
>    Visible links
>    1. https://github.com/massiveart/MassiveSearchBundle/pull/77#discussion_r53626284
>    2. https://github.com/massiveart/MassiveSearchBundle/pull/77/files#r53626284
>    3. https://github.com/massiveart/MassiveSearchBundle/pull/77/files#r53626284